wallet-api/src/Wallet/Emulator/Types.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Wallet.Emulator.Types where
+
+import           Control.Monad.Except
+import           Control.Monad.Operational as Op
+import           Control.Monad.State
+import           Control.Monad.Writer
+import           Data.Either
+import           Data.Map                  as Map
+import           Data.Maybe
+import           Data.Text                 as T
+
+-- Basic types (should be replaced by "real" types later)
+
+-- don’t care what these are, some data to allow typeclasses
+data Txn = Txn Int
+    deriving (Show, Eq, Ord)
+-- each agent has its own wallet state
+data Agent = Agent Int
+    deriving (Show, Eq, Ord)
+
+type Block = [Txn]
+type Chain = [Block]
+type TxPool = [Txn]
+
+data Notification = BlockValidated Block
+                  | BlockHeight Int
+                  deriving (Show, Eq, Ord)
+
+-- Wallet code
+
+data WalletState = WalletState
+    deriving (Show, Eq, Ord)
+
+emptyWalletState :: WalletState
+emptyWalletState = WalletState
+
+-- manually records the list of transactions to be submitted
+newtype EmulatedWalletApi a = EmulatedWalletApi { runEmulatedWalletApi :: StateT WalletState (Writer [Txn]) a }
+    deriving (Functor, Applicative, Monad, MonadState WalletState, MonadWriter [Txn])
+
+-- TODO
+instance WalletAPI EmulatedWalletApi where
+    submitTxn txn = tell [txn]
+    handleNotifications = undefined
+
+-- TODO
+class WalletAPI m where
+    submitTxn :: Txn -> m ()
+    handleNotifications :: [Notification] -> m ()
```

Comment:
Is this to allow the Plutus client (user of wallet API) to register for notifications? If so then I think it should be possible to register a callback

Reply:
No, this is supposed to be the function that gets called on the wallet when some notifications come in for it. At which point it should add any appropriate blocks to its internal state, trigger reactions, etc.

I didn't write in any of the functions for registering handlers etc, but obviously those should be there.

Reply:
I see, I thought that we would eventually have two implementations of `WalletAPI`, namely the mock wallet API and the "real" wallet API. In the real one it wouldn't make much sense to notify the wallet like this (because it would receive its notifications from the core node, which would not be visible to the client).

Reply:
Maybe remove it from the class and write

```
handleNotifications :: MonadEmulator m => [Notification] -> m ()
handleNotifications = ...
```

Reply:
Ah I see. So the version here should just work on the emulated wallet.